Update PHPunit docs (#1085)
shopware/docs: guides/plugins/plugins/testing/php-unit.md

## Overview

This guide will cover the creation of PHPUnit tests in Shopware 6. Please visit the official PHPUnit documentation for a deep dive into PHP unit testing.
<!-- markdown-link-check-disable-next-line -->
{% embed url="https://phpunit.de/documentation.html" caption="" %}

## Prerequisites

> In order to create tests for a plugin, you need a plugin as a base. Please refer to the [Plugin Base Guide](../plugin-base-guide.md).

Reviewer: ```suggestion
In order to create tests for a plugin, you need a plugin as a base. Refer to the [Plugin Base Guide](../plugin-base-guide.md) for more information.
```

## PHPUnit configuration

> First of all we have to configure PHPUnit a bit. Therefore we have to create a file called `phpunit.xml` in the root directory of our plugin. To get more familiar with the configurable options, you can refer to the [PHPUnit documentation](https://phpunit.readthedocs.io/en/8.5/configuration.html). In this example we configure PHPUnit to search in the directories `<plugin root>/src/Test` and `<plugin root>/src/Migration/Test` for our tests.

Reviewer: ```suggestion
First, to configure PHPUnit, create a file called `phpunit.xml` in the root directory of your plugin. To get more familiar with the configurable options, refer to the [PHPUnit documentation](https://phpunit.readthedocs.io/en/8.5/configuration.html). This example explains configuring PHPUnit to search in the directories `<plugin root>/src/Test` and `<plugin root>/src/Migration/Test` for your tests.
```

> The file below is automatically generated when using the `bin/console plugin:create` command

Reviewer: ```suggestion
The file below is automatically generated when using the `bin/console plugin:create` command.
```

> After we've configured PHPUnit, we can start writing our first test. In this example we have a test which simply tries to instantiate every `.php` class, to see if any used core classes went missing. In our test we use the `IntegrationTestBehaviour` trait which comes in with some handy features, such as automatically setting up a database transaction or clearing the cache before starting your tests.

Reviewer: ```suggestion
After having configured PHPUnit, you can start writing your first test. In this example, you have a test that simply tries to instantiate every `.php` class to see if any used core classes are missing. In your test, you use the `IntegrationTestBehaviour` trait, which comes with some handy features, such as automatically setting up a database transaction or clearing the cache before starting your tests.
```

### Migration test

> In order to test our example migration `Migration1611740369ExampleDescription`, we create a new test called `Migration1611740369ExampleDescriptionTest` which extends from the PHPUnit `TestCase`. We use the `KernelTestBehaviour` trait since we need our database connection from the container.

Reviewer: ```suggestion
In order to test the example migration `Migration1611740369ExampleDescription`, create a new test called `Migration1611740369ExampleDescriptionTest`, which extends from the PHPUnit `TestCase`. Use the `KernelTestBehaviour` trait since you need the database connection from the container.
```

In this test-only service config you can override arguments, aliases or parameters to change what the service container injects into your services during a test run.

## Executing the test
To execute tests, you need a PHPUnit binary, which is most likely located in your `vendor/bin` folder. The command below will use the `phpunit.xml` file in the `custom/plugins/SwagBasicExample` folder and execute the testsuite with the name migration.

{% code title="<project root>" %}

```shell
./vendor/bin/phpunit --configuration="custom/plugins/SwagBasicExample" --testsuite "migration"
```

{% endcode %}

### Executing all tests in the plugin

> If no testsuite is passed it will execute all testsuites.

Reviewer: ```suggestion
If no testsuite is passed, it will execute all testsuites.
```

Now we can execute the `phpunit.sh` to run all tests of our plugin.

```text

> ./bin/phpunit.sh

Reviewer 2: do we really need this wrapper script? I would rather show the real phpunit command with the options. Feels weird to wrap it especially in the docs.

Author: I guess it was added to remove one point of error since passing the wrong path will break the execution. But to understand the command better, adding the full command makes sense :+1:

```

### Executing a single test

If we want to execute a specific test only, we have to pass the path to the test as argument.

> ```text

Reviewer 3: If you remove the leading `$` as indicator for a user/non-root command, wouldn't it be a good idea to mark these sections as shell scripts so highlighting suggest, what it could be instead?

Author: @JoshuaBehrens yes, they will be marked as ```shell in the future.

We discussed this in the team and the benefit of easy copying overweighed the indicator benefit. In 99% of the cases the usage is clear (bin/console etc.)

Reviewer 3: I am with you. It is really annoying to remove these leading indicators :) just wanted to know, whether you will update the codefence hint as well
